src/hotspot/share/classfile/classFileParser.cpp

```diff
@@ -4728,13 +4728,19 @@ void ClassFileParser::verify_legal_class_modifiers(jint flags, const char* name,
   const bool is_enum       = (flags & JVM_ACC_ENUM)       != 0;
   const bool is_annotation = (flags & JVM_ACC_ANNOTATION) != 0;
   const bool major_gte_1_5 = _major_version >= JAVA_1_5_VERSION;
+  const bool valid_value_class = is_identity || is_interface ||
+                                 (supports_inline_types() && (!is_identity && (is_abstract || is_final)));
 
   if ((is_abstract && is_final) ||
       (is_interface && !is_abstract) ||
       (is_interface && major_gte_1_5 && (is_identity || is_enum)) ||   //  ACC_SUPER (now ACC_IDENTITY) was illegal for interfaces
-      (!is_interface && major_gte_1_5 && is_annotation)) {
+      (!is_interface && major_gte_1_5 && is_annotation) ||
+      (!valid_value_class)) {
     ResourceMark rm(THREAD);
     const char* class_note = "";
+    if (!valid_value_class) {
+      class_note = " (a value class must be final or else abstract)";
+    }
     if (!is_identity)  class_note = " (a value class)";
```

Comment:
@fparain you needed to delete line 4744 when incorporating my changes. Now you have overwritten the detailed message. I can fix that in my upcoming changes.